setuptools/command/easy_install.py

```diff
@@ -1786,13 +1788,18 @@ def _first_line_re():
     return re.compile(first_line_re.pattern.decode())
 
 
-def auto_chmod(func, arg, exc):
-    if func in [os.unlink, os.remove] and os.name == 'nt':
-        chmod(arg, stat.S_IWRITE)
-        return func(arg)
-    et, ev, _ = sys.exc_info()
-    # TODO: This code doesn't make sense. What is it trying to do?
-    raise (ev[0], ev[1] + (" %s %s" % (func, arg)))
+# Must match shutil._OnExcCallback
+def auto_chmod(func: Callable[..., _T], arg: str, exc: BaseException) -> _T:
+    """shutils onexc callback to automatically call chmod for certain functions."""
+    if os.name != 'nt':
+        raise OSError(f"Can't call auto_chmod on non-nt os {os.name!r}") from exc
+    supported_methods = {os.unlink, os.remove}
+    if func not in supported_methods:
+        raise ValueError(
+            f"Argument func is not one of {[method.__name__ for method in supported_methods]} (got: {func.__name__!r})"
+        ) from exc
+    chmod(arg, stat.S_IWRITE)
+    return func(arg)
```

Comment:
Alternatively, less detailed, but much more concise:
```suggestion
    if func in [os.unlink, os.remove] and os.name == 'nt':
        chmod(arg, stat.S_IWRITE)
        return func(arg)
    raise exc
```